stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+    }
+
+    #[cfg(test)]
+    pub fn num_pending(&self) -> usize {
+        if self.reply.is_some() {
+            1
+        } else {
+            0
+        }
+    }
+
+    /// Set up the pending response
+    /// Called indirectly from ProtocolFamily::read_preamble() when handling an HTTP response
+    /// Used for dealing with streaming data
+    fn set_pending(&mut self, preamble: &HttpResponsePreamble) {
+        self.reply = Some(StacksHttpReplyData {
+            request_id: preamble
+                .get_request_id()
+                .unwrap_or(HTTP_REQUEST_ID_RESERVED),
+            stream: StacksHttpRecvStream::new(MAX_MESSAGE_LEN as u64),
+        });
+    }
+
+    /// Set the preamble. This is only relevant for receiving an HTTP response to a request that we
+    /// already sent.  It gets called from ProtocolFamily::read_preamble().
+    ///
+    /// This method will set up this state machine to consume the message associated with this
+    /// premable, if the response is chunked.
+    fn set_preamble(&mut self, preamble: &StacksHttpPreamble) -> Result<(), NetError> {
+        match preamble {
+            StacksHttpPreamble::Response(ref http_response_preamble) => {
+                // we can only receive a response if we're expecting it
+                if self.request_handler_index.is_none() {
+                    return Err(NetError::DeserializeError(
+                        "Unexpected HTTP response: no active request handler".to_string(),
+                    ));
+                }
+                if http_response_preamble.is_chunked() {
+                    // we can only receive one response at a time
+                    if self.reply.is_some() {
+                        test_debug!("Have pending reply already");
+                        return Err(NetError::InProgress);
+                    }
+
+                    self.set_pending(http_response_preamble);
+                }
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    /// Clear any pending response state -- i.e. due to a failed request.
+    fn reset(&mut self) -> () {
+        self.request_handler_index = None;
+        self.reply = None;
+    }
+
+    /// Used for processing chunk-encoded streams.
+    /// Given the preamble and a Read, stream the bytes into a chunk-decoder.  Return the decoded
+    /// bytes if we decode an entire stream.  Always return the number of bytes consumed.
+    /// Returns Ok((Some(decoded bytes we got, total number of encoded bytes), number of bytes gotten in this call)) if we're done decoding.
+    /// Returns Ok((None, number of bytes gotten in this call)) if there's more to decode.
+    pub fn consume_data<R: Read>(
+        &mut self,
+        preamble: &HttpResponsePreamble,
+        fd: &mut R,
+    ) -> Result<(Option<(Vec<u8>, usize)>, usize), NetError> {
+        assert!(preamble.is_chunked());
+        assert!(self.reply.is_some());
+
+        if let Some(reply) = self.reply.as_mut() {
```

Comment:
Just a preference, but I rather be explicit with something like this (remove the unreachable branch)
```rust 
       let reply = self.reply.as_mut().ok_or_else(|| NetError::CustomError(
            "Empty reply. Nothing to consume".into(),
        ))?;
```

Reply:
Fixed